cupy/cuda/texture.pyx

```diff
@@ -0,0 +1,304 @@
+        self.depth = depth
+        self.flags = flags
+        self.ndim = 3 if depth > 0 else 2 if height > 0 else 1
+
+    def __dealloc__(self):
+        if self.ptr:
+            runtime.freeArray(self.ptr)
+            self.ptr = 0
+
+    cdef int _get_kind(self, src, dst):
+        cdef int kind
+        if isinstance(src, ndarray) and dst is self:
+            kind = runtime.memcpyDeviceToDevice
+        elif src is self and isinstance(dst, ndarray):
+            kind = runtime.memcpyDeviceToDevice
+        elif isinstance(src, numpy.ndarray) and dst is self:
+            kind = runtime.memcpyHostToDevice
+        elif src is self and isinstance(dst, numpy.ndarray):
+            kind = runtime.memcpyDeviceToHost
+        else:
+            raise
+        return kind
+
+    cdef void* _make_cudaMemcpy3DParms(self, src, dst):
+        '''Private helper for data transfer. Supports all dimensions.'''
+        cdef Memcpy3DParms* param = \
+            <Memcpy3DParms*>PyMem_Malloc(sizeof(Memcpy3DParms))
+        c_memset(param, 0, sizeof(Memcpy3DParms))
+        cdef PitchedPtr srcPitchedPtr, dstPitchedPtr
+        cdef intptr_t ptr
+
+        # get kind
+        param.kind = <MemoryKind>self._get_kind(src, dst)
+
+        # get src
+        if src is self:
+            # Important: cannot convert from src.ptr!
+            param.srcArray = <Array>(self.ptr)
+            param.extent = runtime.make_Extent(self.width, self.height,
+                                               self.depth)
+        else:
+            width = src.shape[-1]
+            if src.ndim >= 2:
+                height = src.shape[-2]
+            else:
+                height = 1  # same "one-stride" trick here
+
+            if isinstance(src, ndarray):
+                ptr = src.data.ptr
+            else:  # numpy.ndarray
+                ptr = src.ctypes.data
+
+            srcPitchedPtr = runtime.make_PitchedPtr(
+                ptr, width*src.dtype.itemsize, width, height)
+            param.srcPtr = srcPitchedPtr
+
+        # get dst
+        if dst is self:
+            # Important: cannot convert from dst.ptr!
+            param.dstArray = <Array>(self.ptr)
+            param.extent = runtime.make_Extent(self.width, self.height,
+                                               self.depth)
+        else:
+            width = dst.shape[-1]
+            if dst.ndim >= 2:
+                height = dst.shape[-2]
+            else:
+                height = 1  # same "one-stride" trick here
+
+            if isinstance(dst, ndarray):
+                ptr = dst.data.ptr
+            else:  # numpy.ndarray
+                ptr = dst.ctypes.data
+
+            dstPitchedPtr = runtime.make_PitchedPtr(
+                ptr, width*dst.dtype.itemsize, width, height)
+            param.dstPtr = dstPitchedPtr
+
+        return <void*>param
+
+    def _prepare_copy(self, arr, stream, direction):
+        # sanity checks:
+        # - check shape
+        if self.ndim == 3:
```

Comment:
Since you know the dims, instead of making tricks to make the 3d copy work in every case
try to use the specialized copies for each case, it might solve the issue ...

Reply:
This would be our last resort. In fact I've done that for 2D, see d1e5fed6. I removed it because it's harder to maintain. 

My current guess is that something internal to CuPy is conflicting with `CUDAArray`, since the error didn't show up when we don't do too many cupy operations and focus solely on testing `Test2DTexture`. In addition, the fact that `TestCUDAArray` always succeeds shows that the copy functions are working correctly.